File: Z_1.go
package main

import (
 "fmt"
 "math"
)

const E = 0.0001

func f(x float64) float64 {
 return x*math.Log(x+1) - 1
}
//1 производ
func f1(x float64) float64 {
 return x/(x+1) + math.Log(x+1)
}

func halfDivision(a, b float64) {
 if f(a)*f(b) >= 0.0 {
  fmt.Println("Неверный интервал")
  return
 }
 k := 0
 var c float64
 for math.Abs(b-a) >= E {
  fmt.Printf("%7d|%8.4f | %8.4f|%f\n", k, a, b, b-a)
  c = (a + b) / 2
  if f(c) == 0.0 {
   break
  } else if f(c)*f(a) < 0 {
   b = c
  } else {
   a = c
  }
  k++
 }

 fmt.Printf("Корень %f с %d итерациями\n", c, k)
}

func newtonMethod(a, b int, x0 float64) {
 if f(float64(a))*f(float64(b)) >= 0.0 {
  fmt.Println("Неверный интервал ")
  return
 }
 k := 0
 x1 := x0 - f(x0)/f1(x0)
 for math.Abs(x1-x0) >= E {
  fmt.Printf("%7d|%8.4f | %8.4f|%f\n", k, x0, x1, x1-x0)
  x0 = x1
  x1 = x0 - f(x0)/f1(x0)
  k++
 }
 fmt.Printf("Корень %f с %d итерациями\n", x0, k)
}

func simpleIterations(x0 float64) {
 k := 0
 x1 := 1 / math.Log(x0+1)
 for math.Abs(x1-x0) > E {
  fmt.Printf("%7d|%8.4f | %8.4f|%f\n", k, x0, x1, x1-x0)
  x0 = x1
  x1 = 1 / math.Log(x0+1)
  k++
 }
 fmt.Printf("Корень %f с %d итерациями\n", x0, k)
}

func main() {
 fmt.Println("программу уточнения корня методом половинного деления с точностью до E")
 halfDivision(-1, 1)
 fmt.Println("\nпрограмму уточнения корня Ньютона с точностью до E")
 newtonMethod(-10, 10, 10)
 fmt.Println("\nметода простых итераций")
 simpleIterations(2)
}

File: Z_1.cpp
#include <iostream>
#include <math.h>
#include <vector>
#include <iomanip>
#include <Windows.h>
using namespace std;
const double E = 0.0001;
//𝑥 𝑙𝑛( 𝑥 + 1) = 1
double f(double x) {
	return (x * log(x + 1) - 1);
}
double f1(double x) {
	return x / (x + 1) + log(x + 1);
}
//методом половинного деления
int halfDivision(double a, double b, vector<double>& roots) {
	if (f(a) * f(b) >= 0.0) {
	cout << "Неверный интервал" << endl;
		return 0;
	}
	int k = 0;
	double c = 0;
	while (abs(b - a) >= E) {
		cout << setw(7) << k << "|" << setw(8) << fixed << setprecision(4) << a << " | " << setw(8) << b << "|" << b - a << endl;
		c = (a + b) / 2;
		if (f(c) == 0.0) {
			break;
		}
		else if (f(c) * f(a) < 0) {
			b = c;
		}
		else {
			a = c;
		}
		k++;
	}
	roots.push_back(c);
	cout << "Корень " << c << " с " << k << " итерациями" << endl;
	return k;
}
//Метод Ньютона
int newtonMethod(double x0, vector<double>& roots) {
	int k = 0;
	double x1 = x0 - f(x0) / f1(x0);
	while (abs(x1 - x0) >= E) {
		cout << setw(7) << k << "|" << setw(8) << fixed << setprecision(4) << x0 << " | " << setw(8) << x1 << "|" << x1 - x0 << endl;
		x0 = x1;
		x1 = x0 - f(x0) / f1(x0); // формула для x
		k++;
	}
	roots.push_back(x0);
	cout << "Корень " << x0 << " с " << k << " итерациями" << endl;
	return k;
}
//метода простых итераций
int simpleIterations(double x0, vector<double>& roots) {
	int k = 0;
	double x1 = 1 / log(x0 + 1);
	while (abs(x1 - x0) > E) {
		cout << setw(7) << k << "|" << setw(8) << fixed << setprecision(4) << x0 << " | " << setw(8) << x1 << "|" << x1 - x0 << endl;
		x0 = x1;
		x1 = 1 / log(x0 + 1);
		k++;
	}
	roots.push_back(x0);
	cout << "Корень " << x0 << " с " << k << " итерациями" << endl;
	return k;
}

int main() {
	SetConsoleCP(1251);
	SetConsoleOutputCP(1251);
	cout << "программу уточнения корня методом половинного деления с точностью до E" << endl;
	vector<double> roots;
	int s1 = halfDivision(-1, 1, roots);
	cout << endl << "программу уточнения корня Ньютоном с точностью до E" << endl;
	int s2 = newtonMethod(10, roots);
	cout << endl << "метода простых итераций" << endl;
	int s3 = simpleIterations(2, roots);
	cout << endl << "все корни уравнения" << endl;
	for (double i : roots) {
		cout << i << " ";
	}
	cout << endl << "Скорость сходимости: " << endl;
	cout << endl << "У Ньютона  " << s2 << " итераций" << endl;
	cout << endl << "У метода простых итераций  " << s3 << " итераций" << endl;
	cout << endl << "У  методом половинного деления " << s1 << " итераций" << endl;
}

